pkg/storage/secret/metadata/outbox_store.go

```diff
@@ -0,0 +1,111 @@
+package metadata
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/grafana/grafana/pkg/apis/secret/v0alpha1"
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/registry/apis/secret/assert"
+	"github.com/grafana/grafana/pkg/registry/apis/secret/contracts"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/storage/secret/migrator"
+)
+
+// Implements contracts.OutboxQueue
+type outboxStore struct {
+	db db.DB
+}
+
+func ProvideOutboxQueue(db db.DB) contracts.OutboxQueue {
+	return &outboxStore{db: db}
+}
+
+type outboxMessageDB struct {
+	MessageID       string                      `xorm:"pk 'uid'"`
+	MessageType     contracts.OutboxMessageType `xorm:"message_type"`
+	Name            string                      `xorm:"name"`
+	Namespace       string                      `xorm:"namespace"`
+	EncryptedSecret string                      `xorm:"encrypted_secret"`
+	KeeperName      string                      `xorm:"keeper_name"`
+	ExternalID      *string                     `xorm:"external_id"`
+	Created         int64                       `xorm:"created"`
+}
+
+func (*outboxMessageDB) TableName() string {
+	return migrator.TableNameSecureValueOutbox
+}
+
+func (s *outboxStore) Append(ctx context.Context, input contracts.AppendOutboxMessage) (string, error) {
+	assert.True(input.Type > 0, "outboxStore.Append: outbox message type is required")
+
+	var messageID string
+	err := s.db.InTransaction(ctx, func(ctx context.Context) error {
+		return s.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+			outboxDB := outboxMessageDB{
+				MessageID:   uuid.New().String(),
+				MessageType: input.Type,
+				Name:        input.Name,
+				Namespace:   input.Namespace,
+				KeeperName:  input.KeeperName,
+				ExternalID:  input.ExternalID,
+				Created:     time.Now().UTC().UnixMilli(),
+			}
+			if input.Type == contracts.CreateSecretOutboxMessage || input.Type == contracts.UpdateSecretOutboxMessage {
+				outboxDB.EncryptedSecret = input.EncryptedSecret.DangerouslyExposeAndConsumeValue()
+			}
+			_, err := sess.Table(migrator.TableNameSecureValueOutbox).Insert(outboxDB)
+			if err != nil {
+				return fmt.Errorf("inserting message into secure value outbox table: %+w", err)
+			}
+
+			messageID = outboxDB.MessageID
+			return nil
+		})
+	})
+	return messageID, err
+}
+
+func (s *outboxStore) ReceiveN(ctx context.Context, n uint) ([]contracts.OutboxMessage, error) {
+	messages := make([]contracts.OutboxMessage, 0)
+	err := s.db.InTransaction(ctx, func(ctx context.Context) error {
+		return s.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+			rows := make([]outboxMessageDB, 0)
+			// TODO: skip locked rows
```

Comment:
Do we need a `locked` BOOL column for that?

Reply:
We could just use skip locked if not for sqlite